File: src/main/java/net/greenjab/fixedminecraft/registry/block/OxidizableRailBlock.java
package net.greenjab.fixedminecraft.registry.block;

import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.block.WeatheringCopper;
import net.minecraft.world.level.block.state.BlockState;
import org.jspecify.annotations.NonNull;

public class OxidizableRailBlock extends CopperRailBlock implements WeatheringCopper {

    private final WeatheringCopper.WeatherState weatherState;

    public OxidizableRailBlock(WeatheringCopper.WeatherState weatherState, Properties settings) {
        super(weatherState, settings);
        this.weatherState = weatherState;
    }

    @Override
    public boolean isRandomlyTicking(BlockState state) {
        return WeatheringCopper.getNext(state.getBlock()).isPresent();
    }

    @Override
    protected void randomTick(@NonNull BlockState state, @NonNull ServerLevel world, @NonNull BlockPos pos, @NonNull RandomSource random) {
    this.changeOverTime(state, world, pos, random);}

    @Override
    public @NonNull WeatherState getAge() {
        return this.weatherState;
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/block/OxidizableRailBlock.kt
package net.greenjab.fixedminecraft.registry.block

import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.block.BlockState
import net.minecraft.block.Oxidizable
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.random.Random

class OxidizableRailBlock(internal val oxidation: Oxidizable.OxidationLevel, settings: Settings) : CopperRailBlock(settings), Oxidizable {
    override fun getCodec(): MapCodec<OxidizableRailBlock> = CODEC
    override fun hasRandomTicks(state: BlockState) = Oxidizable.getIncreasedOxidationBlock(state.block).isPresent
    override fun getDegradationLevel() = oxidation
    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) =
        tickDegradation(state, world, pos, random)

    companion object {
        val CODEC: MapCodec<OxidizableRailBlock> = RecordCodecBuilder.mapCodec {
            it.group(
                Oxidizable.OxidationLevel.CODEC.fieldOf("weathering_state")
                    .forGetter(OxidizableRailBlock::getDegradationLevel),
                createSettingsCodec()
            ).apply(it, ::OxidizableRailBlock)
        }
    }
}
